packages/workers-shared/asset-worker/src/handler.ts

```diff
@@ -1,40 +1,125 @@
 import {
+	FoundResponse,
 	InternalServerErrorResponse,
 	MethodNotAllowedResponse,
+	MovedPermanentlyResponse,
+	NoIntentResponse,
 	NotFoundResponse,
 	NotModifiedResponse,
 	OkResponse,
+	PermanentRedirectResponse,
+	SeeOtherResponse,
 	TemporaryRedirectResponse,
 } from "../../utils/responses";
-import { getHeaders } from "./utils/headers";
+import { attachCustomHeaders, getAssetHeaders } from "./utils/headers";
+import { generateRulesMatcher, replacer } from "./utils/rules-engine";
 import type { AssetConfig } from "../../utils/types";
 import type EntrypointType from "./index";
 import type { Env } from "./index";
 
-export const handleRequest = async (
+const REDIRECTS_VERSION = 1;
+export const HEADERS_VERSION = 2;
+
+type AssetIntent = {
+	eTag: string;
+	status: typeof OkResponse.status | typeof NotFoundResponse.status;
+};
+
+const getResponseOrAssetIntent = async (
 	request: Request,
 	env: Env,
 	configuration: Required<AssetConfig>,
-	exists: typeof EntrypointType.prototype.unstable_exists,
-	getByETag: typeof EntrypointType.prototype.unstable_getByETag
+	exists: typeof EntrypointType.prototype.unstable_exists
 ) => {
-	const { pathname, search } = new URL(request.url);
+	const url = new URL(request.url);
+	const { host, search } = url;
+	let { pathname } = url;
+
+	const staticRedirectsMatcher = () => {
```

Comment:
The additions here seem very similar to the `generateResponse()` function in pages-shared. I assume you considered sharing that logic between the two as well? Why did you not?

Reply:
I moved the ahead-of-time parsing and config, but left the runtime evaluation. Felt right, but I definitely won't die on that hill.

My vague thinking was that if we ever fixed a bug in metafiles, we'd probably only do so in Workers Assets and leave Pages frozen (we have access to compatibility date in Workers Assets!). This split makes it _reasonably_ easy to do that.